web/styles/rendered_markdown.css

```diff
@@ -148,6 +140,24 @@
         padding: 4px;
     }
 
+    /* Zero out the margin-bottom on last children, so that there are no
+       spacing side-effects in the message area originating from Markdown
+       elements. */
+    & blockquote:last-child,
+    & h1:last-child,
+    & h2:last-child,
+    & h3:last-child,
+    & h4:last-child,
+    & h5:last-child,
+    & h6:last-child,
+    & ol:last-child,
+    & p:last-child,
+    & pre:last-child,
+    & table:last-child,
+    & ul:last-child {
+        margin-bottom: 0;
+    }
```

Comment:
This looks overbroad. For example, do you intend for this to apply to `.rendered-markdown > blockquote > blockquote > pre:last-child` (which is the last child of the inner `blockquote`, but not necessarily the last child of the entire message)?

Can we achieve this effect more cleanly with margin collapsing, or with negative padding on the message box itself?

Reply:
It is broad, which I think is arguably one possible way to go. In this broad treatment, we let the containing element be in charge of its own spacing. I think that makes it easier to reason about any given element's footprint, and how distant it will be from another element: I want to think about how far a blockquote is from another element that follows, regardless of which element is the last child of the blockquote, or even something more complicated like a nested situation that you describe.

At the same time, I can also see where arriving at one or two spacing values shared by all elements introduces a level of predictability where we might want narrower, immediate-child selectors (e.g., `.rendered_markdown > blockquote:last-child`) to address only the impact of spacing values on the message-content area itself.

I don't think [negative padding on the message box](https://codepen.io/karlstolley/pen/qBwJZZK) will work for cinching up extra space. And while negative margin could work, I think that's more difficult to reason about, especially because dev tools appear to highlight the space occupied by an element with positive margin--which could be confusing especially to newer contributors:

![Screenshot 2024-04-17 at 11 20 58 AM](https://github.com/zulip/zulip/assets/170719/b33aced2-8f34-4e2e-aebd-9bc2f525bc04)

Reply:
I agree broad selectors like this is asking for trouble wherever we insert arbitrary HTML inside a message. Concretely, I'd be surprised if there's no case where putting these syntax elements inside a blockquote or something results in this CSS applying unintentionally. So we should figure out how to make this specific to the end-of-message scenario it's intended for.

One could also imagine an invisible end-of-message HTML element or pseudo-element that's designed to be invisible but has something like

```
div.invisible-message-content-end-marker {
    margin-top: 3px;
    padding-top: -3px;
    .
}
```

to eat the end-of-message margins for whatever the last element is with collapsing, without actually consuming space? Not sure if anything in that direction is viable.

I guess another approach would be a `ul + div.invisible-message-content-end-marker` selector.

Reply:
Maybe, but I'm dubious about whether tossing another element in the DOM on each and every message is a good way to go...those can end up like well-intentioned `<br />` tags, which become pieces of furniture to have to constantly design around.

Keeping the selector specific to the last element in the rendered markdown area is trivial--it just calls for adding `>` between `&` and the last-child element in question, e.g., `& > p:last-child`.

Reply:
That being said, I'm willing to experiment with any of the ideas here, once I get to the stage where we're drawing on the database-stored line-height values and trying to avoid any remaining unexpected shifts up and down the message feed. It's possible we won't know the exact right call to make here until that point--which I suppose is an argument for just scrapping, at least for now, all of the `:last-child` selectors here and going with a negative `margin-bottom` on the message feed area. That'll make for a cleaner PR and way less code to adjust later on when a more optimal solution has a chance to present itself.